src/IO/External/InterpolateFromCocal.cpp

```diff
@@ -0,0 +1,204 @@
+
+  // Pass the string and its length to Fortran
+  const char* dir_cstr =
+      data_directory.c_str();  // Convert std::string to C string
+  int len_dir = static_cast<int>(
+      data_directory.length());  // Get the length of the string
+
+  if (id_type == CocalIdType::Co) {
+    coc2cac_co(num_points, x_coords.data(), y_coords.data(), z_coords.data(),
+               dir_cstr, len_dir, lapse.data(), shift_x.data(), shift_y.data(),
+               shift_z.data(), spatial_metric_xx.data(),
+               spatial_metric_xy.data(), spatial_metric_xz.data(),
+               spatial_metric_yy.data(), spatial_metric_yz.data(),
+               spatial_metric_zz.data(), extrinsic_curvature_xx.data(),
+               extrinsic_curvature_xy.data(), extrinsic_curvature_xz.data(),
+               extrinsic_curvature_yy.data(), extrinsic_curvature_yz.data(),
+               extrinsic_curvature_zz.data(), rest_mass_density.data(),
+               spatial_velocity_x.data(), spatial_velocity_y.data(),
+               spatial_velocity_z.data(), pressure.data(),
+               specific_internal_energy.data());
+  } else if (id_type == CocalIdType::Ir) {
+    coc2cac_ir(num_points, x_coords.data(), y_coords.data(), z_coords.data(),
+               dir_cstr, len_dir, lapse.data(), shift_x.data(), shift_y.data(),
+               shift_z.data(), spatial_metric_xx.data(),
+               spatial_metric_xy.data(), spatial_metric_xz.data(),
+               spatial_metric_yy.data(), spatial_metric_yz.data(),
+               spatial_metric_zz.data(), extrinsic_curvature_xx.data(),
+               extrinsic_curvature_xy.data(), extrinsic_curvature_xz.data(),
+               extrinsic_curvature_yy.data(), extrinsic_curvature_yz.data(),
+               extrinsic_curvature_zz.data(), rest_mass_density.data(),
+               spatial_velocity_x.data(), spatial_velocity_y.data(),
+               spatial_velocity_z.data(), pressure.data(),
+               specific_internal_energy.data());
+  } else if (id_type == CocalIdType::Sp) {
+    coc2cac_sp(num_points, x_coords.data(), y_coords.data(), z_coords.data(),
+               dir_cstr, len_dir, lapse.data(), shift_x.data(), shift_y.data(),
+               shift_z.data(), spatial_metric_xx.data(),
+               spatial_metric_xy.data(), spatial_metric_xz.data(),
+               spatial_metric_yy.data(), spatial_metric_yz.data(),
+               spatial_metric_zz.data(), extrinsic_curvature_xx.data(),
+               extrinsic_curvature_xy.data(), extrinsic_curvature_xz.data(),
+               extrinsic_curvature_yy.data(), extrinsic_curvature_yz.data(),
+               extrinsic_curvature_zz.data(), rest_mass_density.data(),
+               spatial_velocity_x.data(), spatial_velocity_y.data(),
+               spatial_velocity_z.data(), pressure.data(),
+               specific_internal_energy.data());
+  }
+
+  get(get<gr::Tags::Lapse<DataVector>>(result)) = to_datavector(lapse);
```

Comment:
I'd just create `DataVector`s above then pass their `.data()` member's results pointer to COCAL then or even use `get<gr::Tags::Lapse<DataVector>>(result).data()` in the COCAL call (though on will have to check that this already has the correct size).